vcd/resource_vcd_nsxt_edgegateway_l2_vpn_tunnel.go

```diff
@@ -0,0 +1,394 @@
+				Optional: true,
+				Computed: true,
+				Description: "Network CIDR block over which the session interfaces. Only relevant if " +
+					"`session_mode` is set to `SERVER`",
+				ValidateFunc: validation.IsCIDR,
+			},
+			"connector_initiation_mode": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "Connector initation mode of the session describing how a connection is made. " +
+					"Needs to be set only if `session_mode` is set to `SERVER`",
+				ValidateFunc: validation.StringInSlice([]string{"INITIATOR", "RESPOND_ONLY", "ON_DEMAND"}, false),
+			},
+			"pre_shared_key": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: "Pre-shared key used for authentication, needs to be provided only for" +
+					"`SERVER` sessions.",
+			},
+			"peer_code": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				Description: "Base64 encoded string of the full configuration of the tunnel provided by the SERVER session. " +
+					"It is a computed field for SERVER sessions and is a required field for CLIENT sessions.",
+			},
+			"stretched_network": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				// DHCP forwarding supports up to 8 IP addresses
+				Description: "Org VDC networks that are attached to the L2 VPN tunnel",
+				Elem:        stretchedNetwork,
+			},
+		},
+	}
+}
+
+var stretchedNetwork = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"network_id": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "ID of the Org VDC network",
+		},
+		"tunnel_id": {
+			Type:         schema.TypeInt,
+			Optional:     true,
+			Computed:     true,
+			Description:  "Tunnel ID of the network for the tunnel. Read-only for `SERVER` sessions.",
+			ValidateFunc: validation.IntBetween(1, 4093),
+		},
+	},
+}
+
+func resourceVcdNsxtEdgegatewayL2VpnTunnelCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	unlock, err := vcdClient.lockParentVdcGroupOrEdgeGateway(d)
+	if err != nil {
+		return diag.Errorf("[L2 VPN Tunnel create] %s", err)
+	}
+
+	defer unlock()
+
+	orgName := d.Get("org").(string)
+	edgeGatewayId := d.Get("edge_gateway_id").(string)
+
+	nsxtEdge, err := vcdClient.GetNsxtEdgeGatewayById(orgName, edgeGatewayId)
+	if err != nil {
+		return diag.Errorf("[L2 VPN Tunnel create] error retrieving Edge Gateway: %s", err)
+	}
+
+	tunnelConfig, err := readL2VpnTunnelFromSchema(d)
+	if err != nil {
+		return diag.Errorf("[L2 VPN Tunnel create] %s", err)
+	}
+	tunnel, err := nsxtEdge.CreateL2VpnTunnel(tunnelConfig)
+	if err != nil {
+		return diag.Errorf("[L2 VPN Tunnel create] error creating L2 VPN Tunnel: %s", err)
+	}
+	d.SetId(tunnel.NsxtL2VpnTunnel.ID)
+
+	return resourceVcdNsxtEdgegatewayL2VpnTunnelRead(ctx, d, meta)
+}
+
+func resourceVcdNsxtEdgegatewayL2VpnTunnelRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```

Comment:
Just throwing the idea, could this function be merged with the data source read?
I see they're quite similar, the only difference is that DS reads by name first. Maybe you can wrap this difference with an `if origin == "datasource"` (or check by ID first if it is there, otherwise by name).

This came up as it is done throughout the code in other resources.

Reply:
True, managed to wrap it into an elegant-ish solution